### PR TITLE
Fix duplicate bases error (MROs) with GenericAlias

### LINKED ISSUE #905: Duplicates found in MROs (OrderedSet)
### Steps to reproduce
This issue is directly linked to PyCQA/pylint#4093

1. Write the following module:
```
from ordered_set import OrderedSet

class RegistryOrderedSet:
    def __init__(self, clazz, type_name="Object"):
        self._objects_dict = defaultdict(OrderedSet)
```
2. Lint it

### Current behavior
```
Traceback (most recent call last):
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/decorators.py", line 34, in cached
    return cache[func]
KeyError: <bound method ClassDef.slots of <ClassDef.OrderedSet l.55 at 0x7f2517b98dc0>>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/peillexg/PROGRAMMATION/Python/pylint_sandbox/bug_pylint_4093/test/bin/pylint", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/peillexg/PROGRAMMATION/Python/pylint/bin/pylint", line 5, in <module>
    run_pylint()
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/__init__.py", line 22, in run_pylint
    PylintRun(sys.argv[1:])
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/run.py", line 358, in __init__
    linter.check(args)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 862, in check
    self._check_files(
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 896, in _check_files
    self._check_file(get_ast, check_astroid_module, name, filepath, modname)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 922, in _check_file
    check_astroid_module(ast_node)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 1054, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 1099, in _check_astroid_module
    walker.walk(ast_node)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/checkers/variables.py", line 1182, in visit_importfrom
    module = node.do_import_module(name_parts[0])
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/mixins.py", line 99, in do_import_module
    return mymodule.import_module(
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 646, in import_module
    return MANAGER.ast_from_module_name(absmodname)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/manager.py", line 191, in ast_from_module_name
    return self.ast_from_file(found_spec.location, modname, fallback=False)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/manager.py", line 100, in ast_from_file
    return AstroidBuilder(self).file_build(filepath, modname)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 139, in file_build
    return self._post_build(module, encoding)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 159, in _post_build
    self.delayed_assattr(delayed)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 235, in delayed_assattr
    if not _can_assign_attr(inferred, node.attrname):
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 60, in _can_assign_attr
    slots = node.slots()
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/decorators.py", line 36, in cached
    cache[func] = result = func(*args, **kwargs)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2838, in slots
    slots = list(grouped_slots())
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2823, in grouped_slots
    for cls in self.mro()[:-1]:
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2909, in mro
    return self._compute_mro(context=context)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2898, in _compute_mro
    unmerged_mro = list(clean_duplicates_mro(unmerged_mro, self, context))
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 110, in clean_duplicates_mro
    raise exceptions.DuplicateBasesError(
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (OrderedSet), (_GenericAlias, _Final, object), (_GenericAlias, _Final, object), (_GenericAlias, _GenericAlias) for <ClassDef.OrderedSet l.55 at 0x7f2517b98dc0>
```

### Expected behavior
No error. It was the case with `astroid 2.4.2`. A bisection leads to commit 7f1d5134d89aab9b3e49030676637debfd731dbb.

### ``python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"`` output
2.5

### PR DESCRIPTION
**Pull request by cdce8p**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description
Fix duplicate bases error that occurred for Python `<3.9` when using multiple inheritance from `_GenericAlias`. Starting with `3.9` this is parsed differently as `_SpecialGenericAlias` and `_BaseGenericAlias`.

--

As this is a bug fix, it would be great if it could be included in the next release `2.5.1`.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|   | :sparkles: New feature |
|   | :hammer: Refactoring  |
|   | :scroll: Docs |

## Related Issue
Closes #905

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]

File: astroid/scoped_nodes.py [hunk 1]
```diff
@@ -105,7 +105,12 @@ def clean_duplicates_mro(sequences, cls, context):
             (node.lineno, node.qname()) if node.name else None for node in sequence
         ]
         last_index = dict(map(reversed, enumerate(names)))
-        if names and names[0] is not None and last_index[names[0]] != 0:
+        if (
+            names
+            and names[0] is not None
+            and last_index[names[0]] != 0
+            and names[0][1] != "typing._GenericAlias"
+        ):
```
Comment (hippo91) [inline astroid/scoped_nodes.py:113]: I'm not found of this kind of solution. I think there is maybe something to handle through TypingBrain instead.
However i'm not sure, i'm still investigating this.
By the way on my machine the bug is also present with python 3.9.

**Review by hippo91 — COMMENTED** [review]
@cdce8p thanks for this review. 
However i would prefer we wait a little bit to merge it because i think there is probably a better way to handle this.
I would like to dig in a bit more.

File: astroid/scoped_nodes.py (hunk 1 above)
Comment (cdce8p) [reply astroid/scoped_nodes.py:113]: I understand. Thanks for your work!

> By the way on my machine the bug is also present with python 3.9.

Hmm, interesting. MacOS with Py 3.9.2 works. Please let me know if I can help test things, once you have found a better solution.

**Comment by cdce8p** [comment]
I took another look at it and even tried to modify `brain_typing.py`: https://github.com/PyCQA/astroid/compare/master...cdce8p:brain_typing-extension
That worked. However I quickly noticed that we would have to add all generic aliases from the `typing` module to fix all issues which would be a lot of work.

After some more investigation, I found that pylint normally would raise a `duplicate-bases` (E0241) error in such cases. There is even an existing false-positive issue https://github.com/PyCQA/pylint/issues/2717. Since this wasn't the case, it meant that the `astroid.DuplicateBasesError` was raised unexpectedly.

That's when I though about adding a flag to guard the exception during the ast parsing which I've done in the last commit.
This resolves the issue and only leaves the false-positive for E0241 which can be ignored by pylint.

--

A good test example is:
```py
from typing import Sized, Hashable
class Derived(Sized, Hashable):
    def __init__(self):
        self.var = 1
```

**Comment by hippo91** [comment]
@cdce8p  i opened #913 that try to fix the same issue.
IMHO it is more generic than what you proposes here.
I would greatly appreciate your opinion on it, as the one of @AWhetter and @Pierre-Sassoulas.
It still failing on `pypy` and `python3.6`. I will investigate it ASAP.

**Comment by cdce8p** [comment]
> @cdce8p i opened #913 that try to fix the same issue.
> IMHO it is more generic than what you proposes here.
> I would greatly appreciate your opinion on it, as the one of @AWhetter and @Pierre-Sassoulas.
> It still failing on `pypy` and `python3.6`. I will investigate it ASAP.

@hippo91 I just saw it. On first glance, your solution does look quite elegant (and way better than mine). I'll run some tests on my code to see if the issue is completely fixed and will get back to you.

**Comment by cdce8p** [comment]
Replaced by #913

**Pull request closed without merging**